src/lib/kdf/sp800_108/sp800_108.cpp

```diff
@@ -22,51 +23,37 @@ std::unique_ptr<KDF> SP800_108_Counter::new_object() const {
    return std::make_unique<SP800_108_Counter>(m_prf->new_object());
 }
 
-void SP800_108_Counter::kdf(uint8_t key[],
-                            size_t key_len,
-                            const uint8_t secret[],
-                            size_t secret_len,
-                            const uint8_t salt[],
-                            size_t salt_len,
-                            const uint8_t label[],
-                            size_t label_len) const {
-   const std::size_t prf_len = m_prf->output_length();
-
-   const uint64_t blocks_required = (key_len + prf_len - 1) / prf_len;
-
-   if(blocks_required > 0xFFFFFFFF) {
-      throw Invalid_Argument("SP800_108_Counter output size too large");
+void SP800_108_Counter::perform_kdf(std::span<uint8_t> key,
+                                    std::span<const uint8_t> secret,
+                                    std::span<const uint8_t> salt,
+                                    std::span<const uint8_t> label) const {
+   if(key.empty()) {
+      return;
    }
 
-   const uint8_t delim = 0;
-   const uint32_t length = static_cast<uint32_t>(key_len * 8);
+   const auto length = static_cast<uint32_t>(key.size() * 8);
+   const auto prf_len = m_prf->output_length();
 
-   uint8_t* p = key;
-   uint32_t counter = 1;
-   uint8_t be_len[4] = {0};
-   secure_vector<uint8_t> tmp;
+   const uint64_t blocks_required = (key.size() + prf_len - 1) / prf_len;
+   BOTAN_ARG_CHECK(blocks_required <= 0xFFFFFFFF, "SP800_108_Counter output size too large");
```

Comment:
Given that the counter starts at 1, this should probably be:

```suggestion
   BOTAN_ARG_CHECK(blocks_required < 0xFFFFFFFE, "SP800_108_Counter output size too large");
```